<img src="./images/aims-za-logo.jpeg" alt="drawing" style="width:400px;"/>
<h1 style="text-align: center;"><a title="EMS-AIMS-ZA-2024-25" href="https://evansdoe.github.io/aims-za/ems/2024-25">Experimental Mathematics Using SageMath — AIMS-ZA-2024-25</a></h1>


## Instructors: 


* <a href="http://evansdoe.github.io">**Evans Ocansey**</a>

## Day 06 — New Experiments: Sum of Squares and Plotting <a class="anchor" id="new-experiment-sum-of-squares-and-plotting"></a>

[comment]: <> (<h2 style="text-align: left;">Day 02 — Introduction to <a title="SageMath"href="http://www.sagemath.org/"><em>SageMath</em></a>: A Mathematics Software for All</h2>)




The outline of the this notebook is as follows:

## Table of Contents: <a class="anchor" id="day-05-toc"></a> 
* [ ] [<font color=blue>Reviewing the Frobenius Problem</font>](#reviewing-the-frobenius-problem)
  * [<font color=blue>Last Thoughts on the Conductor</font>](#last-thoughts-on-the-conductor)
* [ ] [<font color=blue>New Experiment — Sum of Squares </font>](#new-experiment-sum-of-squares)
  * [<font color=blue>Exploratory Questions — Sum of Squares </font>](#exploratory-questions-sum-of-squares)
* [ ] [<font color=blue>More Plotting and Graphics in Two Dimensions</font>](#more-plotting-and-graphics-in-two-dimensions)

## Reviewing the Frobenius Problem <a class = "anchor" id="reviewing-the-frobenius-problem"></a>
Our motivational problem is the following: 
 - Given positive integers $a$ and $b$, what numbers $m$ can be written in the form $a\,x+b\,y$, where $x,\,y \in \mathbb{Z}_{\ge0}$? 

Given positive integers $a$ and $b$ if there is a positive integer $c$, with $c = a\,x + b\,y$ such that for all positive integers $n \ge c$, there are non-negative integers $x$ and $y$ with $n = a\,x + b\,y$, then $c$ is called the *conductor* of $a$ and $b$. In other words, $c_{\{a,b\}}$ is the smallest positive integer that can be written as $a\,x + b\,y$ such that  all positive integers greater than or equal to $c_{\{a,b\}}$ are also writable. We will denote the conductor of $a$ and $b$ by $c_{\{a,b\}}$.  
On the other hand, the largest positive integer $f$ that can not be written as $f = a\,x + b\,y$ is called the  *Frobenius number* of $a$ and $b$. We will denote this by $f_{\{a,b\}}$.

It has already led to several related questions. Hopefully you were able to explore the following questions and made interesting _conjectures_.

 - Given two positive integers $a$  and $b$, how many numbers can not be written as $a\,x + b\,y$ with $x,\,y\in\mathbb{Z}_{\ge0}$?
 - If $a$ is even and $b$ is odd, what is the Frobenius number? 
 - If $a$ is even and $b$ is even, what is the Frobenius number? 
 - If $a$ is odd and $b$ is odd, what is the Frobenius number? 
 - If $a$ and $b$ are prime numbers, what is the Frobenius number? 
 - Given two positive integers $a$ and $b$, can we find a formula or some other means of computing the Frobenius number?
 - If $a,\, b \in \mathbb{Z}_{>0}$ and $\gcd(a, b) \neq 1$, does the Frobenius number exist?

Here are some questions that we possed together in class: 
- Given that either $a = 1$ or $b = 1$, but not both, is it possible to define a Frobenius number, $f_{\{a,b\}}$ and a conductor, $c_{\{a,b\}}$? If so what is it?
- If $a$ and $b$ have a common factor greater than $1$, does there exist a Frobenius number, $f_{\{a,b\}}$ and a conductor, $c_{\{a,b\}}$?
- If $a$ is even and $b$ is odd what is the Frobenius number and conductor?
- If $f_{\{a,b\}}=\infty$ for $a,\,b \in \mathbb{Z}_{>0}$, is it always true that the conductor, $c_{\{a,b\}}$ is also infinity. How about the vice-versa of this statement?
- Is there a case such that given $a,\,b \in \mathbb{Z}_{>0}$ there exists a Frobenius number, $f_{\{a,b\}}$ but not a conductor, $c_{\{a,b\}}$?
- What is the existence condition for a Frobenius number, $f_{\{a,b\}}$ and a conductor, $c_{\{a,b\}}$ given $a, b\in\mathbb{Z}_{>0}$?
- If $a,b\in\mathbb{Z}_{>0}$ and there exist a Fronbenius number, $f_{\{a,b\}}$ and a conductor, $c_{\{a,b\}}$, is there a relationship between:
  - $f_{\{a,\,b\}}$ and $c_{\{a,\,b\}}$?
  - $f_{\{a,\,b\}}$, $a$, and $b$?
  - $c_{\{a,\,b\}}$, $a$, and $b$?

### Last Thoughts on the Conductor <a class = "anchor" id="last-thoughts-on-the-conductor"></a>

Are there more questions we can explore about the Frobenius problem?

- What if instead of two positive integers $a$ and $b$, we have more than $3$ positive integers $a, b, c$?  

**Question**: Is there such a thing as a *conductor* or a *Frobenius number* for a set of _three_ integers?
That is, is there a conductor for numbers of the form $n=5x+8y+7z$, for the set $\{5,8,7\}$?

   * Well, what about for $\{3,4,7\}$?
   * Okay, what about $\{3,4,6\}$?
   * Ah, but what about $\{3,4,5\}$?

**Question**: We observed from our experiments with two positive integers $a$ and $b$ that the existence criterion for the Frobenius number $f_{\{a,b\}}$ is that $a$ and $b$ must be coprime, i.e., $\gcd(a, b)=1$. Would this criterion also hold for the case of three positive integers: $a$, $b$, and $c$?


Do you think there a formula for the Frobenius number $f_{\{a,b\}}$, and the conductor $c_{\{a,b\}}$?  What about for bigger sets of integers, for example, for three positive integers $a$, $b$, and $c$?

   * For sets of two, it turns out there is a _complete answer_ which some of you have either discovered it already or busy discovering.
   * For sets of three numbers, like $n=5\,x+8\,y+7\,z$, there is an efficient algorithm, but no exact formula.

Notice the connection between theory and practice!  Solving this problem could easily solve a "real-life" business problem, so a good algorithm would be very important. But finding such an algorithm requires exploration and experimentation like we have been doing.  

This [video](https://www.youtube.com/watch?v=vNTSugyS038) gives us a practical application of the Frobenius problem in real life.

Click [**here**](#day-05-toc) if you want to go back to the table of content of this notebook. Otherwise, continue to the next [**Section**](#new-experiment-sum-of-squares).

## New Experiment — Sum of Squares <a class="anchor" id="new-experiment-sum-of-squares"></a>

With that in mind, let's begin our next exploration. Hopefully you have had some fun experimenting with the condunctor problem. Today we will introduce a new problem to keep up the fun spirit of experimental mathematics. Namely, what numbers can be written in the form $n=a^2+b^2$?

Here are a cases that do and do not work.

   * $2 = 1^2+1^2$ works;
   * $3$ does not work, since we only allow integers.  $1+1$ is too small, $1^2+2^2=5$ is too big;
   * $4$ actually <em>does</em> work, because we allow $4 = 0^2+2^2$.  (That's just one of our rules; you could play a different game if you wanted.);
   * That means $1=1^2+0^2$ works.
   * $5$ works.
And so forth. 

## Exploratory Questions — Sum of Squares <a class="anchor" id="exploratory-questions-sum-of-squares"></a>
Can you think of questions that you can explore experimentally? Last Friday, we formulated the following questions. I also added extra questions.

1. Which positive integers can be written as a sum of two squares?
2. Which ones cannot be written?
3. Are numbers that can be represented as a sum of two squares always odd or always even? Or do they vary?
4. What are the sum of squares that are even or odd integers?
5. Is the sum of squares of two numbers generally even or odd? What factors influence this?
6. What are the sum of squares that are perfect squares? What is the probability that the sum of squares is a perfect square? When is this condition true?
7. If $a$ is a factor of $b$, under what conditions is the sum of their squares a perfect square?
8. Can either $a$ or $b$ be a factor of the sum of their squares?
9. For which values of $a$ and $b$ is the sum of their squares a prime number?
10. Can there be two or more distinct pairs of $a$ and $b$ that yield the same sum of squares? If so, what are these pairs?
11. Is the sum of the squares of two consecutive integers always odd? If not, when does it differ?
12. If $a$ and $b$ are both odd, is the sum of their squares odd or even?
13. If $a$ is odd and $b$ is even, what is the sum of their squares?
14. If $a$ is the only even prime (i.e., $2$) and $b$ is an odd prime, is the sum of their squares a prime number?
15. Which numbers can be written as a sum of squares of two prime numbers?
16. Which positive integers can be expressed as the sum of squares of their factors? Are there identifiable patterns?
17. Given $a$ and $b$ are both prime, under what conditions is $a^2 + b^2$ also prime?
18. Are there forms other than $x^2 + 0^2$ for representing a square root as a sum of squares?

Each question could reveal patterns or conjectures. Let's proceed systematically as follows:

   - **Identify** when the statement holds true or false.
   - **Analyze patterns** in the results generated.
   - **Make conjectures** about underlying rules or relationships based on observed patterns.

- Let us start by working with pens and paper.
- I will create a table on the board for us to fill it out.
- We will divide ourselves into groups and try the case for $n$ running from $1$ to $100$ to fillout the table.
- We will study the table to see if we can observe some patterns in the data and make some interesting conjectures.
- When we are done, we will go to the computer lab and see how we can translate the table we have created using an appropriate Python datatype.
- Once we have the tabular data correctly encoded, we will further explore our questions in groups using SageMath.

Hopefully you have tried some examples. Now let us have ourselves in groups and explore our questions using _SageMath_. As you explore, try to write your observation. Later on I will call on the spokesperson of a group to report their findings.

In [1]:
# sorted([a^2 + b^2 for a in [0..10] for b in [0..10]])

In [2]:
def sum_of_squares_data(a_bound=6, b_bound=6):
    sum_of_squares_database = {}
    for a in [0..a_bound]:
        for b in [0..b_bound]:
            n = a^2 + b^2
            if n not in sum_of_squares_database:
                sum_of_squares_database[n] = [(a, b)]
            else:
                sum_of_squares_database[n].append((a,b))
    return sum_of_squares_database

In [3]:
a_bound=20
b_bound=20

sum_of_squares_database = sum_of_squares_data(a_bound=a_bound, b_bound=b_bound)
sum_of_squares_database

{0: [(0, 0)],
 1: [(0, 1), (1, 0)],
 4: [(0, 2), (2, 0)],
 9: [(0, 3), (3, 0)],
 16: [(0, 4), (4, 0)],
 25: [(0, 5), (3, 4), (4, 3), (5, 0)],
 36: [(0, 6), (6, 0)],
 49: [(0, 7), (7, 0)],
 64: [(0, 8), (8, 0)],
 81: [(0, 9), (9, 0)],
 100: [(0, 10), (6, 8), (8, 6), (10, 0)],
 121: [(0, 11), (11, 0)],
 144: [(0, 12), (12, 0)],
 169: [(0, 13), (5, 12), (12, 5), (13, 0)],
 196: [(0, 14), (14, 0)],
 225: [(0, 15), (9, 12), (12, 9), (15, 0)],
 256: [(0, 16), (16, 0)],
 289: [(0, 17), (8, 15), (15, 8), (17, 0)],
 324: [(0, 18), (18, 0)],
 361: [(0, 19), (19, 0)],
 400: [(0, 20), (12, 16), (16, 12), (20, 0)],
 2: [(1, 1)],
 5: [(1, 2), (2, 1)],
 10: [(1, 3), (3, 1)],
 17: [(1, 4), (4, 1)],
 26: [(1, 5), (5, 1)],
 37: [(1, 6), (6, 1)],
 50: [(1, 7), (5, 5), (7, 1)],
 65: [(1, 8), (4, 7), (7, 4), (8, 1)],
 82: [(1, 9), (9, 1)],
 101: [(1, 10), (10, 1)],
 122: [(1, 11), (11, 1)],
 145: [(1, 12), (8, 9), (9, 8), (12, 1)],
 170: [(1, 13), (7, 11), (11, 7), (13, 1)],
 197: [(1, 14), (14, 1)],
 226:

Conjectures made by students in today's class after working with pens and papers to generate data.

- If $n$ is odd, then the pair $(a,b)$ has one to be even and the other odd.
- If $a$ and $b$ are both even, then $n$ is even and if $a$ and $b$ are both odd, then $n$ is even.
- If $n=a^2+b^2$ for $a,b\in\mathbb{N}$ then $a,b\leq\sqrt{n}$.
- If $n=a^2+b^2$ for $a,b\in\mathbb{N}$, then the maximum number of pairs $(a,b)$ for fixed $n$ is equal to 4.
- If $n$ has 4 pairs of solution, then $n$ is a multiple of $5$.
- If $n$ is prime and $>2$, the number of $(a,b)$ is always 2.
- If $a=b$, $n=2a^2$ or $n=2b^2$. If a or b is equal to zero, then $n=a^2$ or $b^2$.
- If $n$ is a perfect square, the number of $(a,b)$ is 2.
- If $n$ is an odd prime and can be written as 1 mod 4, then it can be written as $a^2+b^2$.

## Conjecture 1 

### - If $n$ is odd, then the pair $(a,b)$ has one to be even and the other odd.

### Experiment 1 <a id="experiment-01"></a>

In [4]:
exp1 ={key: sum_of_squares_database[key] for key in list(filter(is_odd, sum_of_squares_database))}

In [5]:
# Define the odd_even lambda function to check if the tuple has one odd and one even number
odd_even = lambda tup: (is_odd(tup[0]) and is_even(tup[1])) or (is_even(tup[0]) and is_odd(tup[1]))
all_odd_even = all(all(odd_even(tup) for tup in value) for value in exp1.values())

print(f"All tuples have one odd and one even number which is: {all_odd_even}")


All tuples have one odd and one even number which is: True


# Conjecture 2

# If $a$ and $b$ are both even, then $n$ is even and if $a$ and $b$ are both odd, then $n$ is even.

## This is a trivial question the square of all odd numbers is always even
* Then given $a$ and $b$ are all odd their sum is even (sum of two odds is even)
## The sum of two evens squared is always even
* Then given $a$ and $b$ are all even their sum is even (sum of two even is even)

# Conjecture 3

# If $n=a^2+b^2$ for $a,b\in\mathbb{N}$ then $a,b\leq\sqrt{n}$.

In [6]:
# Lambda function to check if both elements of the tuple are less than the square root of the key
ab_less_sqrt_n = lambda tup, key: (tup[0] <= sqrt(key)) and (tup[1] <= sqrt(key))

# Check all values in sum_of_squares_database
all(all(ab_less_sqrt_n(tup, key) for tup in value) for key, value in sum_of_squares_database.items())

True

## This conjecture is True for all the vales of $a,b$

# Conjecture 4

# If $n=a^2+b^2$ for $a,b\in\mathbb{N}$, then the maximum number of pairs $(a,b)$ for fixed $n$ is equal to 4.

In [7]:
pairs = {n: pairs for n, pairs in sum_of_squares_database.items() if len(pairs)>4}

In [8]:
pairs

{325: [(1, 18), (6, 17), (10, 15), (15, 10), (17, 6), (18, 1)],
 425: [(5, 20), (8, 19), (13, 16), (16, 13), (19, 8), (20, 5)]}

### This conjecture is not true since we can deprove it by a counter example showing pairs that are actually above 4

# Conjecture 5

# If $n$ has 4 pairs of solution, then $n$ is a multiple of $5$.

In [9]:
pairs_four = {n: pairs for n, pairs in sum_of_squares_database.items() if len(pairs)==4}

In [10]:
pairs_four

{25: [(0, 5), (3, 4), (4, 3), (5, 0)],
 100: [(0, 10), (6, 8), (8, 6), (10, 0)],
 169: [(0, 13), (5, 12), (12, 5), (13, 0)],
 225: [(0, 15), (9, 12), (12, 9), (15, 0)],
 289: [(0, 17), (8, 15), (15, 8), (17, 0)],
 400: [(0, 20), (12, 16), (16, 12), (20, 0)],
 65: [(1, 8), (4, 7), (7, 4), (8, 1)],
 145: [(1, 12), (8, 9), (9, 8), (12, 1)],
 170: [(1, 13), (7, 11), (11, 7), (13, 1)],
 290: [(1, 17), (11, 13), (13, 11), (17, 1)],
 85: [(2, 9), (6, 7), (7, 6), (9, 2)],
 125: [(2, 11), (5, 10), (10, 5), (11, 2)],
 260: [(2, 16), (8, 14), (14, 8), (16, 2)],
 365: [(2, 19), (13, 14), (14, 13), (19, 2)],
 130: [(3, 11), (7, 9), (9, 7), (11, 3)],
 205: [(3, 14), (6, 13), (13, 6), (14, 3)],
 265: [(3, 16), (11, 12), (12, 11), (16, 3)],
 370: [(3, 19), (9, 17), (17, 9), (19, 3)],
 185: [(4, 13), (8, 11), (11, 8), (13, 4)],
 305: [(4, 17), (7, 16), (16, 7), (17, 4)],
 340: [(4, 18), (12, 14), (14, 12), (18, 4)],
 377: [(4, 19), (11, 16), (16, 11), (19, 4)],
 221: [(5, 14), (10, 11), (11, 10), (14, 

In [11]:
divisible_byfive = lambda divisible_byfive: divisible_byfive %5
all(map(divisible_byfive, pairs_four.keys()))

False

## This conjecture is not true as we have some values that are not divisible by 5 ie 221

# Conjecture 6

## If $n$ is prime and $>2$, the number of $(a,b)$ is always 2.

In [12]:
primes = {key: sum_of_squares_database[key] for key in list(filter(is_prime , sum_of_squares_database))}

In [13]:
del primes[2]

In [14]:
len_pair_ab = lambda len_pair_ab: len(len_pair_ab)==2
all(map(len_pair_ab, primes.items()))

True

#### This conjecture is True hence If $n$ is prime and $>2$, the number of $(a,b)$ is always 2.

In [15]:
 n = []
length_n = []
for i in primes.keys():
    n.append(i)
    length_n.append(len(primes[i]))
    

In [16]:
experiment1 = [n,length_n]
header = ["Odd Prime n" , "The Pairs (a,b)"]

table(experiment1, header_column=header)

Odd Prime n     │ 5   17   37   101   197   257   401   13   29   53   173   229   293   73   109   409   41   97   137   241   61   89   281   349   157   397   113   149   193   373   449   233   353   181   277   337   269   389   461   317   521   313   433   569   421   557   617   613   761
  The Pairs (a,b) │ 2   2    2    2     2     2     2     2    2    2    2     2     2     2    2     2     2    2    2     2     2    2    2     2     2     2     2     2     2     2     2     2     2     2     2     2     2     2     2     2     2     2     2     2     2     2     2     2     2

# Conjecture 7

## If $a=b$, $n=2a^2$ or $n=2b^2$. If a or b is equal to zero, then $n=a^2$ or $b^2$.

In [17]:
perfect_n ={key: [tup for tup in value if tup[0] == tup[1]] for key, value in sum_of_squares_database.items()}

In [18]:
# Lambda function to check if both elements of the tuple are less than the square root of the key
perfect = lambda tup, key: (2*tup[0]^2 == key) and (2*tup[1]^2 == key)

# Check all values in sum_of_squares_database
all(all(perfect(tup, key) for tup in value) for key, value in perfect_n.items())

True

## This conjecture is true and was trival result

# Conjecture 8 

## - If $n$ is a perfect square, the number of $(a,b)$ is 2.


In [19]:
perfect_squares ={key: sum_of_squares_database[key] for key in list(sum_of_squares_database) if sqrt(key) == int(key^.5)}; perfect_squares

{0: [(0, 0)],
 1: [(0, 1), (1, 0)],
 4: [(0, 2), (2, 0)],
 9: [(0, 3), (3, 0)],
 16: [(0, 4), (4, 0)],
 25: [(0, 5), (3, 4), (4, 3), (5, 0)],
 36: [(0, 6), (6, 0)],
 49: [(0, 7), (7, 0)],
 64: [(0, 8), (8, 0)],
 81: [(0, 9), (9, 0)],
 100: [(0, 10), (6, 8), (8, 6), (10, 0)],
 121: [(0, 11), (11, 0)],
 144: [(0, 12), (12, 0)],
 169: [(0, 13), (5, 12), (12, 5), (13, 0)],
 196: [(0, 14), (14, 0)],
 225: [(0, 15), (9, 12), (12, 9), (15, 0)],
 256: [(0, 16), (16, 0)],
 289: [(0, 17), (8, 15), (15, 8), (17, 0)],
 324: [(0, 18), (18, 0)],
 361: [(0, 19), (19, 0)],
 400: [(0, 20), (12, 16), (16, 12), (20, 0)],
 625: [(15, 20), (20, 15)]}

In [20]:
length = []
for i in perfect_squares.values():
    length.append(len(i)==2)

In [21]:
all(length)

False

## This conjecture is not true

# Conjecture 9

# If $n$ is an odd prime and can be written as 1 mod 4, then it can be written as $a^2+b^2$.

In [22]:
# %display latex


In [23]:
def oddprime_mudulo_four(primes):
    for i in primes.keys():
        if mod(i,4)==1:
            print( f'  mod({i}, 4) = 1 ')
        else:
            print( f'  mod({i}, 4) != 1 ')


In [24]:
oddprime_mudulo_four(primes)

  mod(5, 4) = 1 
  mod(17, 4) = 1 
  mod(37, 4) = 1 
  mod(101, 4) = 1 
  mod(197, 4) = 1 
  mod(257, 4) = 1 
  mod(401, 4) = 1 
  mod(13, 4) = 1 
  mod(29, 4) = 1 
  mod(53, 4) = 1 
  mod(173, 4) = 1 
  mod(229, 4) = 1 
  mod(293, 4) = 1 
  mod(73, 4) = 1 
  mod(109, 4) = 1 
  mod(409, 4) = 1 
  mod(41, 4) = 1 
  mod(97, 4) = 1 
  mod(137, 4) = 1 
  mod(241, 4) = 1 
  mod(61, 4) = 1 
  mod(89, 4) = 1 
  mod(281, 4) = 1 
  mod(349, 4) = 1 
  mod(157, 4) = 1 
  mod(397, 4) = 1 
  mod(113, 4) = 1 
  mod(149, 4) = 1 
  mod(193, 4) = 1 
  mod(373, 4) = 1 
  mod(449, 4) = 1 
  mod(233, 4) = 1 
  mod(353, 4) = 1 
  mod(181, 4) = 1 
  mod(277, 4) = 1 
  mod(337, 4) = 1 
  mod(269, 4) = 1 
  mod(389, 4) = 1 
  mod(461, 4) = 1 
  mod(317, 4) = 1 
  mod(521, 4) = 1 
  mod(313, 4) = 1 
  mod(433, 4) = 1 
  mod(569, 4) = 1 
  mod(421, 4) = 1 
  mod(557, 4) = 1 
  mod(617, 4) = 1 
  mod(613, 4) = 1 
  mod(761, 4) = 1 


In [25]:
modulo = []
n= []
for i in primes.keys():
    modulo.append(mod(i,4)==1)
    n.append(i)

In [26]:
all(modulo)

True

In [27]:
experiment2= [n, modulo]

In [28]:
header2 = ["Odd Prime n" , "Mod(n,4)"]

table(experiment2, header_column=header2)

Odd Prime n │ 5      17     37     101    197    257    401    13     29     53     173    229    293    73     109    409    41     97     137    241    61     89     281    349    157    397    113    149    193    373    449    233    353    181    277    337    269    389    461    317    521    313    433    569    421    557    617    613    761
  Mod(n,4)    │ True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True   True